### PR TITLE
[K32W] Add NXP BLE Adaptation Layer

### PR DESCRIPTION
**Pull request by doru91**
Add BLE Adaptation Layer for NXP K32W061 platform. Using this PR, K32W can be used for BLE Rendezvous. The Commissioning Parameters are then used for joining a Thread Network.

Note that for K32W we need to manually update OT repo to the latest commit (included documentation in the README). A future PR should update the OT version used by PCHIP - I tried to update the OT stack but then EFR32 stops compiling (L.E.: talked to @jepenven-silabs and the update of OT stack for EFR32 is on their agenda).

Also, the latest SDK release for K32W is SDK2.6.2. In order to fix the compilation for Project CHIP we are patching this SDK with the files pushed to third_party/k32w_sdk. This is just a temporary step until a newer NXP SDK will be released.

### CODE REVIEW

**Comment by doru91** [comment]
The Doxygen error seems to be a false positive:
"/home/runner/work/connectedhomeip/connectedhomeip/src/platform/K32W/ble_function_mux.c:54: error: include file gatt_decl_x.h not found, perhaps you forgot to add its directory to INCLUDE_PATH? (warning treated as error, aborting now)"

gatt_decl_x.h is included in the K32W MR2 SDK.

**Comment by doru91** [comment]
> The Doxygen error seems to be a false positive:
> "/home/runner/work/connectedhomeip/connectedhomeip/src/platform/K32W/ble_function_mux.c:54: error: include file gatt_decl_x.h not found, perhaps you forgot to add its directory to INCLUDE_PATH? (warning treated as error, aborting now)"
> 
> gatt_decl_x.h is included in the K32W MR2 SDK.



> The Doxygen error seems to be a false positive:
> "/home/runner/work/connectedhomeip/connectedhomeip/src/platform/K32W/ble_function_mux.c:54: error: include file gatt_decl_x.h not found, perhaps you forgot to add its directory to INCLUDE_PATH? (warning treated as error, aborting now)"
> 
> gatt_decl_x.h is included in the K32W MR2 SDK.

Found a workaround using the define "DOXYGEN_SHOULD_SKIP_THIS"

**Comment by andy31415** [comment]
> > The Doxygen error seems to be a false positive:
> > "/home/runner/work/connectedhomeip/connectedhomeip/src/platform/K32W/ble_function_mux.c:54: error: include file gatt_decl_x.h not found, perhaps you forgot to add its directory to INCLUDE_PATH? (warning treated as error, aborting now)"
> > gatt_decl_x.h is included in the K32W MR2 SDK.
> 
> > The Doxygen error seems to be a false positive:
> > "/home/runner/work/connectedhomeip/connectedhomeip/src/platform/K32W/ble_function_mux.c:54: error: include file gatt_decl_x.h not found, perhaps you forgot to add its directory to INCLUDE_PATH? (warning treated as error, aborting now)"
> > gatt_decl_x.h is included in the K32W MR2 SDK.
> 
> Found a workaround using the define "DOXYGEN_SHOULD_SKIP_THIS"

Doxyfile also contains some EXCLUDE_PATTERS and EXCLUDE_SYMBOLS if that helps

File: examples/lock-app/k32w/README.md [hunk 1]
```diff
@@ -21,24 +34,49 @@ board.
...
+
+### Bluetooth LE Advertising
+
+To commission the device onto a Project CHIP network, the device must be
+discoverable over Bluetooth LE. Bluetooth LE advertising is started
+automatically when the device is powered up.
+
+### Bluetooth LE Rendezvous
+
+In Project CHIP, the commissioning procedure (called rendezvous) is done over
```
Comment (andy31415) [inline examples/lock-app/k32w/README.md:66]: wording here and above seems to imply that BLE is required, however theoretically CHIP devices do not require BLE (Ethernet-only devices or SoftAP). Should we update to say 'in this example' or 'for thread devices'?

File: examples/platform/k32w/app/ldscripts/chip-k32w061-lock-example.ld [hunk 2]
```diff
@@ -53,25 +53,14 @@ MEMORY
...
- *                - Weave Task = 8192
- *                - Thread Task = 8192
- *                - App Task 4096
- *                - Timer Task: 720
- * - Mbedtls: 8096
- * - Other allocations: 4096
- * TOTAL = 34256
- */
-HEAP_SIZE = DEFINED(HEAP_SIZE) ? HEAP_SIZE : 0x85D0;
+/* first implementation: second RAM bank is dedicated entirely to heap */
```
Comment (andy31415) [inline examples/platform/k32w/app/ldscripts/chip-k32w061-lock-example.ld:62]: Not sure I follow the 'first implementation' comment - is this a left-over or is there some second/subsequent version?

File: src/platform/K32W/BLEManagerImpl.cpp [hunk 3]
```diff
@@ -37,7 +58,1415 @@ namespace Internal {
...
+#define CHIP_DEVICE_CONFIG_BLE_APP_TASK_STACK_SIZE (1024)
+
+/* advertising configuration */
+#define BLEKW_ADV_MAX_NO (2)
+#define BLEKW_SCAN_RSP_MAX_NO (2)
+#define BLEKW_MAX_ADV_DATA_LEN (31)
+#define CHIP_ADV_SHORT_UUID_LEN (2)
+
+/* Message list used to synchronize asynchronous messages from the KW BLE tasks */
+static anchor_t blekw_msg_list;
```
Comment (andy31415) [inline src/platform/K32W/BLEManagerImpl.cpp:109]: nit: keyword 'static' is not required in anonymous namespace

File: src/platform/K32W/BLEManagerImpl.cpp [hunk 4]
```diff
@@ -37,7 +58,1415 @@ namespace Internal {
...
+            return CHIP_ERROR_INVALID_ARGUMENT;
+        }
+        memset(mDeviceName, 0, kMaxDeviceNameLength);
+        strcpy(mDeviceName, deviceName);
+        SetFlag(mFlags, kFlag_DeviceNameSet, true);
+        ChipLogProgress(DeviceLayer, "Setting device name to : \"%s\"", deviceName);
+    }
+    else
+    {
+        mDeviceName[0] = 0;
```
Comment (andy31415) [inline src/platform/K32W/BLEManagerImpl.cpp:366]: Above sets the `kFlag_DevicewNameSet` to true.
Should this set it to false?

File: src/platform/K32W/BLEManagerImpl.cpp [hunk 5]
```diff
@@ -37,7 +58,1415 @@ namespace Internal {
...
+    case DeviceEventType::kCHIPoBLESubscribe: {
+        ChipDeviceEvent connEstEvent;
+
+        HandleSubscribeReceived(event->CHIPoBLESubscribe.ConId, &CHIP_BLE_SVC_ID, &ChipUUID_CHIPoBLEChar_TX);
+        connEstEvent.Type = DeviceEventType::kCHIPoBLEConnectionEstablished;
+        PlatformMgr().PostEvent(&connEstEvent);
+        break;
+    }
+
+    case DeviceEventType::kCHIPoBLEUnsubscribe: {
```
Comment (andy31415) [inline src/platform/K32W/BLEManagerImpl.cpp:385]: nit: in many switch cases, the extra {} do not seem needed (no local variables)

File: src/platform/K32W/BLEManagerImpl.cpp [hunk 6]
```diff
@@ -37,7 +58,1415 @@ namespace Internal {
...
+    {
+        ChipLogProgress(DeviceLayer, "Started Advertising.");
+    }
+    else
+    {
+        ChipLogProgress(DeviceLayer, "Advertising error!");
+        return CHIP_ERROR_INCORRECT_STATE;
+    }
+
+exit:
```
Comment (andy31415) [inline src/platform/K32W/BLEManagerImpl.cpp:841]: is the exit label used anywhere?

This function seems to return right away and not use the exit label pattern. I would suggest using only one approach (either return anywhere, or have an exit lable, not both).

File: src/platform/K32W/BLEManagerImpl.cpp [hunk 7]
```diff
@@ -37,7 +58,1415 @@ namespace Internal {
...
+            else if (msg->type == BLE_KW_MSG_ATT_WRITTEN || msg->type == BLE_KW_MSG_ATT_LONG_WRITTEN ||
+                     msg->type == BLE_KW_MSG_ATT_CCCD_WRITTEN)
+            {
+                sInstance.HandleWriteEvent(msg);
+            }
+            else if (msg->type == BLE_KW_MSG_ATT_READ)
+            {
+                blekw_att_read_data_t * att_rd_data = (blekw_att_read_data_t *) msg->data.data;
+                ChipLogProgress(DeviceLayer, "Attribute read request(d:%d, h:%d) .", att_rd_data->device_id, att_rd_data->handle);
+                /* TODO: */
```
Comment (andy31415) [inline src/platform/K32W/BLEManagerImpl.cpp:963]: add some description to the TODO? what needs to be done?

File: src/platform/K32W/BLEManagerImpl.cpp [hunk 8]
```diff
@@ -37,7 +58,1415 @@ namespace Internal {
...
+
+/* Called by BLE when a connect is received */
+void BLEManagerImpl::BLE_SignalFromISRCallback(void)
+{
+    /* TODO: Low Power */
+}
+
+void BLEManagerImpl::blekw_connection_timeout_cb(TimerHandle_t timer)
+{
+    (void) blekw_msg_add_u8(BLE_KW_MSG_FORCE_DISCONNECT, (uint8_t) 0);
```
Comment (andy31415) [inline src/platform/K32W/BLEManagerImpl.cpp:1246]: is the uint8_t cast required?

File: src/platform/K32W/DeviceNetworkProvisioningDelegateImpl.cpp [hunk 9]
```diff
@@ -0,0 +1,43 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
```
Comment (andy31415) [inline src/platform/K32W/DeviceNetworkProvisioningDelegateImpl.cpp:3]: 2021

File: src/platform/K32W/ble_function_mux.c [hunk 10]
```diff
@@ -0,0 +1,336 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
```
Comment (andy31415) [inline src/platform/K32W/ble_function_mux.c:3]: 2021 or 2020-2021

File: src/platform/K32W/k32w-chip-mbedtls-config.h [hunk 11]
```diff
@@ -1,6 +1,5 @@
 /*
- *   Copyright (c) 2020 Project CHIP Authors
- *  Copyright (c) 2020, The OpenThread Authors.
+ *  Copyright (c) 2019, The OpenThread Authors.
```
Comment (andy31415) [inline src/platform/K32W/k32w-chip-mbedtls-config.h:2]: Add back copyright?

Also this sems to lose some changes (like pragma once)

**Review by andy31415 — APPROVED** [review]

File: examples/lock-app/k32w/README.md (hunk 1 above)
Comment (doru91) [reply examples/lock-app/k32w/README.md]: Right, changed the wording to "In this example".

File: examples/platform/k32w/app/ldscripts/chip-k32w061-lock-example.ld (hunk 2 above)
Comment (doru91) [reply examples/platform/k32w/app/ldscripts/chip-k32w061-lock-example.ld]: Changed the wording to "To be improved. At this moment the second RAM bank is dedicated entirely to heap."

File: src/platform/K32W/BLEManagerImpl.cpp (hunk 3 above)
Comment (doru91) [reply src/platform/K32W/BLEManagerImpl.cpp]: Thanks, removed the 'static' occurrence in anonymous namespace.

File: src/platform/K32W/BLEManagerImpl.cpp (hunk 4 above)
Comment (doru91) [reply src/platform/K32W/BLEManagerImpl.cpp:366]: Set it to false to be on the safe side.

File: src/platform/K32W/BLEManagerImpl.cpp (hunk 5 above)
Comment (doru91) [reply src/platform/K32W/BLEManagerImpl.cpp]: Removed the extra {} in cases where no local variables are declared.

File: src/platform/K32W/BLEManagerImpl.cpp (hunk 7 above)
Comment (doru91) [reply src/platform/K32W/BLEManagerImpl.cpp]: Removed entirely this branch as it's not exercised for this use-case.

File: src/platform/K32W/BLEManagerImpl.cpp (hunk 8 above)
Comment (doru91) [reply src/platform/K32W/BLEManagerImpl.cpp]: Removed the unneeded cast.

File: src/platform/K32W/DeviceNetworkProvisioningDelegateImpl.cpp (hunk 9 above)
Comment (doru91) [reply src/platform/K32W/DeviceNetworkProvisioningDelegateImpl.cpp]: Changed it to 2021.

File: src/platform/K32W/ble_function_mux.c (hunk 10 above)
Comment (doru91) [reply src/platform/K32W/ble_function_mux.c]: Changed it to 2021.

File: src/platform/K32W/k32w-chip-mbedtls-config.h (hunk 11 above)
Comment (doru91) [reply src/platform/K32W/k32w-chip-mbedtls-config.h]: Added the correct copyright - 2021 Project CHIP Authors.

**Comment by doru91** [comment]
I fixed the review problems. Is there any other change required for having the PR approved?

**Review by rwalker-apple — APPROVED** [review]

File: examples/lock-app/k32w/README.md (hunk 1 above)
Comment (LuDuda) [reply examples/lock-app/k32w/README.md]: @doru91 would you mind to make this change in the original source as well? Thanks.

**Review by LuDuda — APPROVED** [review]

**Comment by woody-apple** [comment]
Looks like this could be merged, @jelderton @saurabhst can we have final reviews?

**Review by saurabhst — APPROVED** [review]

File: examples/lock-app/k32w/README.md (hunk 1 above)
Comment (doru91) [reply examples/lock-app/k32w/README.md]: @LuDuda , sure, operated this changes also in the nrfconnect README.

**Review by saurabhst — APPROVED** [review]

**Review by woody-apple — APPROVED** [review]

**Pull request merged**